src/resources/espnow.cc

```diff
@@ -304,54 +418,26 @@ PRIMITIVE(create) {
   int id = espnow_pool.any();
   if (id == kInvalidEspNow) FAIL(ALREADY_IN_USE);
 
-  if (!init()) {
+  event_queue = xQueueCreate(ESPNOW_EVENT_NUM, sizeof(EspNowEvent));
+  if (!event_queue) {
     espnow_pool.put(id);
     FAIL(MALLOC_FAILED);
   }
 
   EspNowResource* resource = _new EspNowResource(group, id, event_queue);
   if (!resource) {
     espnow_pool.put(id);
+    vQueueDelete(event_queue);
     FAIL(MALLOC_FAILED);
   }
 
-  // TODO(florian): we are leaking the resource and
-  // all the allocated entries (from 'init') if one of the
-  // following calls fails.
-
-
-  // Not clear whether we should keep this call to esp_netif_init.
-  // The lwip thread is supposed to do this (and normally does so).
-  // However, it doesn't seem to be guaranteed.
-  // The code looks safe to be executed multiple times, but it's
-  // not clear whether it is thread-safe...
-  esp_err_t err = esp_netif_init();
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  wifi_init_config_t cfg = WIFI_INIT_CONFIG_DEFAULT();
-  wifi_mode_t wifi_mode = mode == 0 ? WIFI_MODE_STA : WIFI_MODE_AP;
+  // From now on the resource is in charge of all allocations. The
+  // event_queue, too, is now deleted by it.
 
-  err = esp_wifi_init(&cfg);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  err = esp_wifi_set_storage(WIFI_STORAGE_RAM);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  err = esp_wifi_set_mode(wifi_mode);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  err = esp_wifi_start();
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-
-  wifi_interface_t interface = mode == 0 ? WIFI_IF_STA : WIFI_IF_AP;
-  err = esp_wifi_config_espnow_rate(interface, phy_rate);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-
-  err = esp_now_init();
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  err = esp_now_register_send_cb(espnow_send_cb);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  err = esp_now_register_recv_cb(espnow_recv_cb);
-  if (err != ESP_OK) return Primitive::os_error(err, process);
-  if (pmk.length() > 0) {
-    err = esp_now_set_pmk(pmk.address());
-    if (err != ESP_OK) return Primitive::os_error(err, process);
+  Object* init_result = resource->init(process, mode, pmk, phy_rate);
```

Comment:
It would be easier to track the lifetime of the pool and the event queue if they were also allocated from within `init`.

Reply:
The event queue is given to the Resource constructor (and its super), but I could allocate the espnow_pool in the init function.